src/llfs/simulated_log_device_storage.cpp

```diff
@@ -0,0 +1,539 @@
+{
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphemeralState::async_write_some(offset=" << file_offset
+               << ", size=" << data.size() << ")";
+
+  if (this->closed_.load()) {
+    BATT_FORWARD(handler)(Status{batt::StatusCode::kClosed});
+    return;
+  }
+
+  this->on_work_started(ASYNC_WRITE_SOME);
+
+  auto* write_task = new batt::Task{
+      this->get_task_scheduler().schedule_task(),
+      [shared_this = batt::shared_ptr_from(this), file_offset, data,
+       handler = BATT_FORWARD(handler)]() mutable {
+        auto* const this_ = shared_this.get();
+
+        // Write all blocks synchronously.
+        //
+        Status status =
+            this_->multi_block_iop<ConstBuffer, &DurableState::write_block>(file_offset, data);
+
+        auto result = [&] {
+          if (status.ok()) {
+            return StatusOr<i32>{static_cast<i32>(data.size())};
+          }
+          return StatusOr<i32>{status};
+        }();
+
+        this_->queue_
+            .push([handler, result, this_] {
+              auto on_scope_exit = batt::finally([this_] {
+                this_->on_work_finished(ASYNC_WRITE_SOME);
+              });
+              handler(result);
+            })
+            .IgnoreError();
+      },
+      batt::Task::DeferStart{true}};
+
+  write_task->call_when_done([write_task] {
+    delete write_task;
+  });
+
+  write_task->start();
+}
+
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+template <typename BufferT,
+          Status (SimulatedLogDeviceStorage::DurableState::*Op)(u64, i64, const BufferT&)>
+Status SimulatedLogDeviceStorage::EphemeralState::multi_block_iop(i64 offset, BufferT buffer)
+{
+  LLFS_VLOG(1) << "(id=" << this->id_ << ") EphmeralState::multi_block_iop(" << BATT_INSPECT(offset)
+               << "," << BATT_INSPECT(buffer.size())
+               << ") type=" << (std::is_same_v<BufferT, ConstBuffer> ? "write" : "read");
+
+  if (this->closed_.load()) {
+    return batt::StatusCode::kClosed;
+  }
+
+  std::vector<Status> block_status(buffer.size() / sizeof(AlignedBlock));
+  batt::Watch<usize> blocks_remaining(block_status.size());
+  std::atomic<usize> pin_count{block_status.size()};
+
+  if (buffer.size() != block_status.size() * sizeof(AlignedBlock)) {
+    return batt::StatusCode::kInvalidArgument;
+  }
+
+  usize i = 0;
+  while (buffer.size() > 0) {
+    LLFS_VLOG(1) << "(id=" << this->id_ << ") EphmeralState::multi_block_iop()"
+                 << BATT_INSPECT(offset) << BATT_INSPECT(i) << "/" << block_status.size();
+
+    this->simulation_post([this, i, offset, &block_status, &blocks_remaining, &pin_count,
+                           block_buffer = BufferT{buffer.data(), sizeof(AlignedBlock)}] {
+      LLFS_VLOG(1) << "(id=" << this->id_ << ") EphmeralState::multi_block_iop() - activated"
+                   << BATT_INSPECT(offset) << BATT_INSPECT(i) << "/" << block_status.size();
+
+      auto on_scope_exit = batt::finally([&] {
```

Comment:
Yes... `batt::finally` works by saving the function/lambda you pass it in an object which runs that function in its destructor.  `batt::finally` returns this object so the caller must capture it if the desired effect is to defer the "final action" until control leaves the current scope.  `on_scope_exit` is actually being used, because its scope is determining when the lambda runs.